Drastic cleanup

moagrius opened the pull request:
Hey guys - here's the most recent.  The biggest change is that I'm letting the `ThreadPoolExecutor` manage our list of `Futures`, and have extended the `Future`s we're passing to have `WeakReferences` to everything they need to know about.

Also, in the interest of thread-safety, in workers themselves, I'm _never_ referencing anything in the main thread (we were previously) - instead, I'm notifying a `Handler` (which is on the main thread) to dispatch messages, and update UI there.  This let me get rid of all the synchronized lists and I think has greatly improved performance, and thread safety.

This seems pretty solid - take a look at the non-HTTP version (`RealMapTileViewActivity`) as well as the HTTP-one - I set it to use Peter's render-while-panning behavior, and it looks very fast to me. 

Obviously, the other issues we were experiencing (the skia stacktraces, missing tiles, etc) are all corrected as well.

Let me know what you think.

Thanks,
Mike

moagrius committed:
with overriden executor, prior to migration to map associations

moagrius committed:
working, prior to removing Future List

moagrius committed:
future list removed in favor of map.keySet

moagrius committed:
removing logs

moagrius committed:
removing logs

moagrius committed:
updating per review

moagrius committed:
wip prior to restart runnable tasks

moagrius committed:
first draft of managed tasks

moagrius committed:
not working

moagrius committed:
wip

moagrius committed:
working

moagrius committed:
working, wip, before shift to Future list

moagrius committed:
wip

moagrius committed:
prior to loop consolidation

moagrius committed:
not working

moagrius committed:
not working, wip

moagrius committed:
not working

moagrius committed:
not working

moagrius committed:
working

moagrius committed:
working

moagrius committed:
working

moagrius committed:
working

moagrius committed:
working

moagrius committed:
working

moagrius committed:
working

moagrius committed:
working, ready for review

moagrius committed:
working, ready for review, more cleanup

moagrius committed:
working, ready for review, more cleanup

p-lr commented on tileview/src/main/java/com/qozix/tileview/tiles/TileRenderHandler.java:
```diff
@@ -0,0 +1,64 @@
+  @Override
+  public void handleMessage( Message message ) {
+    TileRenderRunnable tileRenderRunnable = (TileRenderRunnable) message.obj;
+    TileCanvasViewGroup tileCanvasViewGroup = tileRenderRunnable.getTileCanvasViewGroup();
```
The `TileCanvasViewGroup` instance is retrieved here, but it is not used in each `tileRenderRunnable`.
I suggest that the handler has its own `TileCanvasViewGroup` weakref, and we don't set a weakref for each `tileRenderRunnable`.

moagrius replied:
sounds reasonable

p-lr commented:
Looks very good!
Good idea for calling a handler inside the run method of each `TileRenderRunnable`, and use the more effective `obtainMessage` method.

p-lr commented on tileview/src/main/java/com/qozix/tileview/tiles/TileRenderRunnable.java:
```diff
@@ -0,0 +1,157 @@
+    Handler handler = getHandler();
+    if( handler != null ) {
+      Message message = handler.obtainMessage( status.getMessageCode(), this );
+      message.sendToTarget();
```
As the current implementation of the handler is to do nothing when `TileRenderHandler.Status` is `INCOMPLETE`, maybe we could save us the call for the handler to `obtainMessage` and `sendToTarget` when this case happens.
But if that rarely happens, i don't think that it's a good idea.

moagrius replied:
good point.  i'd also considered giving the user some callback in case of incomplete or error - right now, as you said, nothing happens on incomplete, and i do cleanup on error if it's an OOM, but that might be better as a user configurable thing

p-lr commented on tileview/src/main/java/com/qozix/tileview/tiles/TileRenderRunnable.java:
```diff
@@ -0,0 +1,157 @@
+    if( context == null ) {
+      return TileRenderHandler.Status.INCOMPLETE;
+    }
+    BitmapProvider bitmapProvider = getBitmapProvider();
```
As the `BitmapProvider` is only accessed from within `TileRenderRunnable` class, i suggest that we don't have a public method `getBitmapProvider` and we directly call here : 

```
if( mBitmapProviderWeakReference == null ) {
  return TileRenderHandler.Status.INCOMPLETE;
}
try {
  tile.generateBitmap( context, mBitmapProviderWeakReference.get() );
} catch( Throwable throwable ) {
...
```

moagrius replied:
that'll throw an error if `get` returns null - `getBitmapProvider` is more of a convenience wrapper for null tests - i don't mind changing it from `public` to `private` or `package`

p-lr commented:
I have tested with my own app and the result is very promising : its as smooth as before (but it can deal with tile on network!)
Also, i have checked that `TileCanvasViewGroup.FAST_RENDER_BUFFER` can be safely raised to 30, even for local tiles.

moagrius commented:
Awesome, thanks for testing that - it seems to me that even local tiles seem to render much faster now - did you see that as well?

I'm going to incorporate your comments, including the FAST_RENDER_BUFFER, and do a little more refactoring, then ask for a merge.

Thanks again

p-lr commented:
Hum, i focused on stress test by panning like crazy on a big map, at different scales and especially at scales like 2^n + epsilon (where tiles are scaled at half their original size).
But yes, saw a better render speed for a given scale.

One common point is that i sometime see a short slow-down, and it always happen during a garbage collection. Used memory increases while panning, and garbage collection happens if used memory becomes too high (that can only be seen on a big map, because on the demo you can't pan far enough).
That isn't an issue after all, because in case of a normal use case, garbage collection can hardly be seen and memory regularly freed. To me, that just means that lower memory consumption will result in overall better performances.

moagrius commented:
GTK.  i might be too aggressive on cleanup - and am open to any suggestions - otherwise that's something i'll probably try to give a closer look at once this is merged.

thanks again.

p-lr commented:
I will think about it too. But as it also concerns the previous implementation, it definitely should be part of another PR (but its a small issue).

moagrius commented:
agreed

bnsantos commented:
nice code @moagrius I was testing and I'm not able to see that bump on memory during onPause or onStop. I tested a little bit here and I was able to only one have a column with white tiles (a moto G 3rd), but I was not able to reproduce again (maybe a network issue).

My suggestion for checking the Tiles would be a unique id for each tile, but I know it's late for that kind of suggestion and your equals method is fine!

Sorry for being late in the discussion but I have too much on my plate right now, can't find time to work on Android.

moagrius commented:
thanks for your feedback @bnsantos, i know what it's like to get buried under work, i appreciate you taking the time.

the tiles actually need to _not_ be unique, since they're generated as `new` during intersection detection, and added to set, so I don't want to replace one if one already exists that a particular row and column (and possibly is rendered, or has a worker spawned, etc)

as i mentioned above, i'll do a few more small tests and tweaks and post back with a request to merge - i promise i won't take too much more of your time after this, we're almost done

thanks again

moagrius committed:
after first round of PR

moagrius committed:
after refactor

moagrius committed:
setting fast render buffer

moagrius commented:
hey guys - this (i hope) is the last iteration, and i'm ready to ask for a merge

@bnsantos - if this looks ok, please merge this branch into _your_ master, then create a PR from your fork to mine, which i will merge into the the main/mine/original repo - after that you should be done and i'll try to not bug you again for a while :)

@peterLaurence - i addressed all comments, and refactored where i could.  with `FAST_RENDER_BUFFER`, maybe it was my imagination, but i think i could see a difference - it was 10, you suggested 30, and i landed on 15, which is basically one drawing frame

i'm pretty pleased with how this turned out, it was really a team effort and i really appreciate you guys

thanks again

moagrius commented:
oh also @peterLaurence i forgot to mention - i bet if we put all large cleanup steps on another thread, we'll really even out those memory spikes...  as you say tho, another issue for another time.

p-lr commented:
@moagrius 
Ok for 15 if you saw a difference :)
As for cleanup in another thread, good idea and it's easy to try. Will do that tomorrow.

p-lr commented:
@moagrius i should bang my head on the wall. I just discovered that from the beginning i did not have `setTransitionsEnabled( false )` like in the demos. That result in a huge difference in behavior, including memory consumption. This is way better with transitions disabled : used memory stays low and there are no performance drop.
Otherwise, cleanup is called very often, and despite running it in another thread i could not avoid the slow down i talked about. Anyway, maybe we will talk about it later.

This is very exiting how TileView is fast now :)

moagrius commented:
yeah transitions are expensive - and they way i have them implemented i don't think there's much we could do to speed them up, unfortunately :/

i agree, we'll talk about cleanup after this current PR is done and released

p-lr commented:
@moagrius Maybe you could create a PR to your own repo by adding it to the remotes of @bnsantos fork you are working from.

moagrius commented:
github has functionality built in that i could PR directly to my master, but since this was originally based off @bnsantos work, i'd like the history to show his contribution

moagrius commented:
i'm going to go ahead and merge this myself, and see if i can get this walked into the original repo soon

p-lr commented:
Yes! Can't wait to see that!
